src/ClientHandle.cpp

```diff
@@ -2515,6 +2521,7 @@ void cClientHandle::SendDisconnect(const AString & a_Reason)
 		LOGD("Sending a DC: \"%s\"", StripColorCodes(a_Reason).c_str());
 		m_Protocol->SendDisconnect(a_Reason);
 		m_HasSentDC = true;
+		m_State = csKicked;
```

Comment:
Why not just shutdown the link here? Postponing until the (next) tick seems needless.

Reply:
The server crashes with a segfault if you do that, so I decided it wasn't the best idea.

Reply:
That's weird, and it means a race condition is probably present even in your code (although probably hard-to-replicate).
Can you investigate into that segfault a bit?

Reply:
I will try to, if I can replicate it.

Reply:
I found the race condition that was causing the segfault - once the client closed the connection after receiving the DC packet, `Destroy()` was called on the client, which would attempt to `Shutdown()` a `nullptr` `m_Link`.

However, if `m_Link` is shut down immediately after `m_Protocol->SendDisconnect()` is called, the kick message doesn't get through to the player - the connection must end before the data in transit gets through.

Reply:
Ah, right, now I remember. The packet sending is done from the client tick thread, so the DC packet is queued here, and then at least one tick is needed for it to be actually sent over the link. So yeah, this code is perfect, then.

Perhaps add a note of this next to this line, so that we know in the future.